src/memory/inventory_manager.rs

```diff
@@ -0,0 +1,101 @@
+use crate::memory::memory_context::MemoryContext;
+use crate::memory::{MemoryManager, MemoryManagerUpdate};
+use crate::state::StateContext;
+use data::prelude::all_items;
+use log::info;
+use memory::game_engine::il2cpp::unity_serializable_dictionary::*;
+use memory::memory_manager::il2cpp::UnityMemoryManager;
+use memory::process::{MemoryError, Process};
+use memory::string::ArrayWString;
+use std::hash::Hash;
+
+impl Default for MemoryManager<InventoryManagerData> {
+    fn default() -> Self {
+        let manager = Self {
+            name: "InventoryManager".to_string(),
+            data: InventoryManagerData::default(),
+            manager: UnityMemoryManager::default(),
+        };
+        info!("Memory: {} Loaded", manager.name);
+        manager
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct InventoryManagerData {
+    // This references QuantityByInventoryItemReference which is a
+    // SerializableDictionary<InventoryItemReference, int> which is
+    // T1 = Guid
+    // T2 = Quantity
+    // In memory you may see the values packed within 0x20 Dictionary 0x18 Entries
+    // With a starting point of 0x20; from there the memory is laid out like:
+    // 0x20 Metadata (unused, for internal stuff)
+    // 0x28 InventoryItemReference
+    //   (first 20 bytes are metadata)
+    //   0x14 - beginning of utf8 GUID
+    // 0x30 Quantity
+    // ... so on until count is met or NULL pointer
+    // These types can theoretically originally have a `next` link for key collisions
+    // so be on the watch for missing items
+    pub items: UnitySerializableDictionary<InventoryItemName, InventoryItemQuantity>,
+}
+
+impl MemoryManagerUpdate for InventoryManagerData {
+    fn update(
+        &mut self,
+        ctx: &StateContext,
+        manager: &mut UnityMemoryManager,
+    ) -> Result<(), MemoryError> {
+        let memory_context = MemoryContext::create(ctx, manager)?;
+
+        self.update_items(&memory_context)?;
+
+        Ok(())
+    }
+}
+
+impl InventoryManagerData {
+    pub fn update_items(&mut self, memory_context: &MemoryContext) -> Result<(), MemoryError> {
+        // If currentInventory != 0x0
+        if let Ok(items_ptr) = memory_context.follow_fields::<u64>(&["ownedInventoryItems"]) {
+            if items_ptr != 0x0 {
+                self.items =
+                    UnitySerializableDictionary::<InventoryItemName, InventoryItemQuantity>::read(
+                        memory_context.process,
+                        items_ptr,
+                    )?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Default, Debug, Eq, PartialEq, Hash)]
+pub struct InventoryItemName(pub String);
+
+#[derive(Default, Debug, PartialEq, Eq, Hash)]
+pub struct InventoryItemQuantity(pub u64);
```

Comment:
Don't mind it being here, but this one probably doesn't need `Hash`, no?

Reply:
It weirdly enough does :P Its just because its a tuple-struct being used as a hash key i think. If it was a normal struct i think it would already have hash.

Reply:
Huh, didn't think that the quantity was used as a key anywhere. As a value it makes sense (but then it wouldn't need `Hash`). Ignore and keep as is, minor point :P